pyiceberg/table/__init__.py

```diff
@@ -1290,6 +1291,19 @@ def snapshot_by_name(self, name: str) -> Optional[Snapshot]:
             return self.snapshot_by_id(ref.snapshot_id)
         return None
 
+    def latest_snapshot_before_timestamp(self, timestamp_ms: int) -> Optional[Snapshot]:
+        """Get the snapshot right before the given timestamp, or None if there is no matching snapshot."""
+        result, prev_timestamp = None, 0
+        if self.metadata.current_snapshot_id is not None:
+            for snapshot in self.current_ancestors():
+                if snapshot and prev_timestamp < snapshot.timestamp_ms < timestamp_ms:
```

Comment:
```suggestion
                if snapshot and prev_timestamp < snapshot.timestamp_ms <= timestamp_ms:
```
I think we could also include the given timestamp, making the method return the latest snapshot up until the given timestamp.

Also, how about implementing this by iterating over `table.history()`? The `snapshot_log` field in metadata contains a list of `snapshot_id` + `timestamp` pair so we do not need to re-generate the ancestors for current snapshot.

Reply:
> I think we could also include the given timestamp, making the method return the latest snapshot up until the given timestamp. 

That was my initial approach, however, I saw the [java implementation](https://github.com/apache/iceberg/blob/795fea9446723cf9c2dbc3cac07883cace8997c9/core/src/main/java/org/apache/iceberg/SetSnapshotOperation.java#L146) for rollback which is not inclusive of the given timestamp. This [implementation](https://github.com/apache/iceberg/blob/795fea9446723cf9c2dbc3cac07883cace8997c9/core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java#L350), which is what you describe, exists but is not used for rollback. So I thought there must be a design decision I am unaware of to keep it not inclusive.
WDYT - should I still change it to be inclusive of the timestamp?

> Also, how about implementing this by iterating over table.history()? The snapshot_log field in metadata contains a list of snapshot_id + timestamp pair so we do not need to re-generate the ancestors for current snapshot.

Ok!

Reply:
That's an interesting finding. I am also not sure the exact decision behind whether inclusive. Although whether to include the exact timestamp does not matter a lot in this use case, I think we may still want to keep the edge case behavior the same as Java: that we do not include the given timestamp when `rollback_to_timestamp`. 

Since this will be a public util method, how about adding another parameter like

- `inclusive`
- `include_given_timstamp`
- ...
that default to `True` and we disable it when using it in `rollback_to_timestamp` implementation?

Reply:
updated to use inclusive parameter!